subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy

```diff
@@ -1071,4 +1071,36 @@ class JavaCompileIntegrationTest extends AbstractPluginIntegrationTest {
         then:
         succeeds("compileJava")
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/18262")
+    @Requires(TestPrecondition.JDK9_OR_LATER)
+    def "should compile sources from source with -sourcepath option for modules"() {
+        given:
+        buildFile << """
+            apply plugin: 'java'
+
+            tasks.register("compileCustomJava", JavaCompile) {
+                destinationDirectory.set(new File(buildDir, "classes/java-custom-path/main"))
+                source = files("src/main/java-custom-path").asFileTree
+                options.sourcepath = files("src/main/java") + files("src/main/java-custom-path")
+                classpath = files()
+            }
+        """
+        file("src/main/java/com/example/SourcePathTest.java") << """
+            package com.example;
+
+            public class SourcePathTest {}
+        """
+        file("src/main/java-custom-path/module-info.java") << """
+            module com.example {
+                exports com.example;
+            }
+        """
+
+        expect:
+        succeeds("compileCustomJava")
+        file("build/classes/java-custom-path/main/module-info.class").exists()
+        // We compile only classes defined in `source`
+        !file("build/classes/java-custom-path/main/com/example/SourcePathTest.class").exists()
```

Comment:
There is something I do not follow. Why is that file not compiled? It should be part of `options.sourcepath`, which is custom set and thus to be respected.
Or did I misunderstand something?

Reply:
This PR makes `options.sourcepath` work the same way as it did before 7.0. I am not sure why `options.sourcepath` worked that way before. As I understand `-sourcepath` it's very similar to JavaCompile `source` input, it can also have jars and zips though.
We can change behavior of it. But I would also like to better understand why user chose `options.sourcepath` over Gradle's `source` in this case.